Implement the BatchClient class with tests and documentation

cmcantalupo opened the pull request:
- Relates to #1962 
- Fixes #1964 
- Requires #1992

cmcantalupo committed:
BatchClient implementation and tests

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

Krebrov001 commented on service/test/BatchClientTest.cpp:
```diff
@@ -0,0 +1,136 @@
+TEST_F(BatchClientTest, read_batch_empty)
+{
+    std::vector<double> result_expect;
+    EXPECT_CALL(*m_batch_status, send_message(_))
```
What is `send_message(_)`?

Krebrov001 commented on service/src/BatchClient.hpp:
```diff
@@ -0,0 +1,128 @@
+            virtual ~BatchClientImp() = default;
+            std::vector<double> read_batch(void) override;
+            void write_batch(std::vector<double> settings) override;
+            void stop_batch(void) override;
```
I would like to add a unit test for `stop_batch()`.

Krebrov001 commented on service/src/BatchClient.cpp:
```diff
@@ -0,0 +1,120 @@
+        }
+        if (settings.size() != (size_t)m_num_control) {
+            throw Exception("BatchClientImp::write_batch(): settings vector length does not match the number of configured controls",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```
I want to add a unit test for `write_batch()` that would trigger the exception to be thrown when `std::vector<double> settings` with wrong number of elements is passed in.

Krebrov001 replied:
Maybe it would be useful to move the `if (m_num_control == 0)` below this check to have the error condition get triggered even if `std::vector<double> settings = {1.2, 2.3}` is passed in, when there are zero controls?

Krebrov001 commented:
In general, I would like to add the doxygen `@throws` to all methods which may throw an exception, either directly or indirectly.
https://www.doxygen.nl/manual/commands.html

Krebrov001 committed:
Added new tests to `BatchClient`, coverage of all the codes.

 - `BatchClientTest.write_batch_wrong_size`
   - what happens when the settings vector length does not match the number of configured controls
 - `BatchClientTest.write_batch_wrong_size_empty`
   - similar test, but the configured controls are empty this time
 - `BatchClientTest.stop_batch`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

dannosliwcd commented on service/src/BatchClient.cpp:
```diff
@@ -0,0 +1,120 @@
+
+    void BatchClientImp::stop_batch(void)
+    {
+        m_batch_status->send_message(BatchStatus::M_MESSAGE_QUIT);
+        m_batch_status->receive_message(BatchStatus::M_MESSAGE_QUIT);
```
Will this result in left-over resources on the server side if a client terminates without calling stop_batch()?

I think you already answered this when reviewing the server side, but I can't remember. Maybe a comment here (or when creating this connection) could help.


Also why does this need to receive a quit message? Is there some subsequent work that fails if the server hasn't yet ack'd the stop request?

cmcantalupo replied:
The geopmd process checks that all PID's that have open sessions are active once per second.  If a PID that has an open session terminates, geopmd will stop any batch server running for that session and restore all control values.  This is done with the callback registration below:

https://github.com/geopm/geopm/blob/dev/service/geopmdpy/service.py#L680-L687

There is a problem though... the close_session() method does not shut down the open batch server as it should...

https://github.com/geopm/geopm/blob/dev/service/geopmdpy/service.py#L419-L428

cmcantalupo replied:
Addressed this issue with the rather large patch added to the next PR, #1998

https://github.com/geopm/geopm/pull/1998/commits/b797fba5e412953875aa8442e837251b7dd84c95

Since this does not impact the BatchClient implementation, this can be merged with the #1998

cmcantalupo replied:
Oh and the QUIT message. We send that message whenever the client has made a request and the batch server is ending.  In the highlighted case, we send back the QUIT message when the client requested it.  If the server is interrupted by geopmd while servicing a  READ or WRITE request from the client, the client will wait for the CONTINUE message, but actually receive the QUIT message.  This will be an error on the client side *unless* the request that got interrupted was the QUIT request...

dannosliwcd reviewed (dismissed):
Just raised two questions. LGTM if those are non-issues.

cmcantalupo committed:
Add a comment about how even the quit request is blocking

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
